docs/configuration.md
# Configuration

_Chat Roulette for Slack_ supports configuration via a JSON config file and/or environment variables. Environment variables take precedence over whatever is in the config file.

See [config.example.json](./examples/config.example.json) for an example of how to modify the configuration of the app using a JSON config file.


#### Bot Config

| Key | Environment Variable | Type | Required | Default Value | Description
| -------- | -------- | -------- | :--------: | -------- | ------
| `auth_token` | `BOT_AUTH_TOKEN` | String | Yes |  | The Slack OAuth2 bot token used by the bot when making API calls to Slack.

###### JSON

```json
{
    "bot": {
        "auth_token": "xoxb-9876543210123-4567778889990-f0A2GclR80dgPZLTUEq5asHm"
    }
}
```

#### Database Config

| Key | Environment Variable | Type | Required | Default Value | Description
| -------- | -------- | -------- | :--------: | -------- | ------
| `url` | `DATABASE_URL` | String | Yes |  | The PostgreSQL connection URL in the form: <br /><br />`postgres://username:password@host:5432/database-name`
| `key` | `DATABASE_ENCRYPTION_KEY` | String | Yes |  | The 32-byte hex-encoded encryption key. This is used to encrypt sensitive data stored in the database. <br /><br />Use `make generate/key` to generate a random key.
| `max_open` | `DATABASE_CONNECTIONS_MAX_OPEN` | Integer | No | `20` | The maximum number of open connections.
| `max_idle` | `DATABASE_CONNECTIONS_MAX_IDLE` | Integer | No | `10` | The maximum number of idle connections.
| `max_lifetime` | `DATABASE_CONNECTIONS_MAX_LIFETIME` | String | No | `60m` | The maximum lifetime of connections. This must be a valid [duration string](https://pkg.go.dev/time#ParseDuration).
| `max_idletime` | `DATABASE_CONNECTIONS_MAX_IDLETIME` | String | No | `15m` | The maximum amount of time a connection can be idle. This must be a valid [duration string](https://pkg.go.dev/time#ParseDuration).

###### JSON
```json
{
    "database": {
        "url": "postgres://postgres:complex-password-here@db.example.com:5432/chat-roulette",
        "encryption": {
            "key": "01234abcde5678901234f62c898cdb592eb3166b56da733e8e798305b0ef6403"
        },
        "connections": {
            "max_open": 50,
            "max_idle": 20,
            "max_lifetime": "2h",
            "max_idletime": "20m"
        }
    }
}
```


#### Server Config

| Key | Environment Variable | Type | Required | Default Value | Description
| -------- | -------- | -------- | -------- | -------- | ------
| `address` | `SERVER_ADDRESS` | String | No | `0.0.0.0` | The address that the HTTP server binds on.
| `port` | `SERVER_ADDRESS` | String | No | `8080` | The TCP port that the HTTP server binds on.
| `client_id` | `SERVER_CLIENT_ID` | String | Yes |  | The Slack OpenID Connect (OIDC) Client ID to support _Sign in with Slack_ for the UI. See [here](https://api.slack.com/authentication/sign-in-with-slack).
| `client_secret` | `SERVER_CLIENT_SECRET` | String | Yes |  | The Slack OIDC Client secret.
| `redirect_url` | `SERVER_REDIRECT_URL` | String | Yes |  | The Slack OIDC redirect URL. Must include the path `/oidc/callback`.
| `secret_key` | `SERVER_SECRET_KEY` | String | Yes |  | The 32-byte hex-encoded secret key used to authenticate cookies.<br /><br />Use `make generate/key` to generate a random key.
| `signing_secret` | `SERVER_SIGNING_SECRET` | String | Yes |  | The Slack-provided secret used to authenticate requests received from Slack.


###### JSON
```json
{
    "server": {
        "client_id": "2518545982190.4321012345789",
        "client_secret": "9f8e7dbc9a5ba2aa4522c7c8eb571f60",
        "redirect_url": "https://www.example.com/oidc/callback",
        "secret_key": "8c4faf836e29d282f2dc7ffdf4ef59c6081e2d8964ba0ac9cd4bc8800021300c",
        "signing_secret": "2773fb7eb76c90f19c0e1504ae1eee4b"
    }
}
```

#### Worker Config

| Key | Environment Variable | Type | Required | Default Value | Description
| -------- | -------- | -------- | -------- | -------- | ------
| `concurrency` | `WORKER_CONCURRENCY` | Integer | No | [runtime.NumCPU()](https://pkg.go.dev/runtime#NumCPU) | The number of concurrent workers to run.<br /><br />This defaults to the number of logical CPUs usable by the current process.

###### JSON
```json
{
    "worker": {
        "concurrency": 2
    }
}
```


#### Tracing Config

| Key | Environment Variable | Type | Required | Default Value | Description
| -------- | -------- | -------- | -------- | -------- | ------
| `enabled` | `TRACING_ENABLED` | Boolean | No | `false`  | Setting this to true enables tracing using [OpenTelemetry](https://opentelemetry.io/).
| `exporter` | `TRACING_EXPORTER` | String | No |  | The tracing exporter to use. This must be set if tracing is enabled. <br /><br />Options: <ul><li>`honeycomb`</li><li>`jaeger`</li></ul>
| `endpoint` | `TRACING_JAEGER_ENDPOINT` | String | No |  | The URL of the Jaeger OTLP HTTP collector. This must be set if the tracing exporter is set to `jaeger`.
| `team` | `TRACING_HONEYCOMB_TEAM` | String | No |  | The [honeycomb.io](https://www.honeycomb.io/) API key. This must be set if the tracing exporter is set to `honeycomb`.
| `dataset` | `TRACING_HONEYCOMB_DATASET` | String | No |  | The dataset to send traces to. This must be set if the tracing exporter is set to `honeycomb`.
| - | `OTEL_TRACES_SAMPLER` | String | No | `always_on` | Configure the sampling strategy to be used.<br /><br />Options: <ul><li>`always_on`</li><li>`traceidratio`</li><li>`parentbased_traceidratio`</li></ul><br />Refer to: [General SDK Configuration](https://opentelemetry.io/docs/languages/sdk-configuration/general/#otel_traces_sampler)
| - | `OTEL_TRACES_SAMPLER_ARG` | String | No |  | Configure additional arguments for the sampler<br />Refer to: [General SDK Configuration](https://opentelemetry.io/docs/languages/sdk-configuration/general/#otel_traces_sampler_arg)

###### JSON
```json
{
    "tracing": {
        "enabled": true,
        "exporter": "honeycomb",
        "honeycomb": {
            "team": "abcdef01234567899876543210a1b3c4",
            "dataset": "chat-roulette"
        }
    }
}
```

```json
{
    "tracing": {
        "enabled": true,
        "exporter": "jaeger",
        "jaeger": {
            "endpoint": "http://localhost:4318/v1/traces"
        }
    }
}
```


#### Misc Config

| Key | Environment Variable | Type | Required | Default Value | Description
| -------- | -------- | -------- | -------- | -------- | ------
| `dev` | `DEV` | Boolean | No | `false` | Set to `true` to enable development mode. <br /><br />Development mode disables verifying requests received from Slack. This should not be enabled for production deployments.

###### JSON
```json
{
    "dev": false
}
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html),
and is generated by [Changie](https://github.com/miniscruff/changie).

## v1.3.1 - 2024-10-04

### Fixed
* Twitter social links now allow x.com domain - [#52](https://github.com/chat-roulettte/chat-roulette/pull/51)
* Trace sampling can be enabled for Jaeger endpoint using environment variables - [#52](https://github.com/chat-roulettte/chat-roulette/pull/52)
## v1.3.0 - 2024-09-30

### Added
* Added user option to match only with others of the same gender - [#34](https://github.com/chat-roulettte/chat-roulette/pull/34)
* Bot will notify participants who did not get matched - [#35](https://github.com/chat-roulettte/chat-roulette/pull/35)

### Changed
* The first round of Chat Roulette will not begin until the user who invites the bot to the channel completes onboarding - [#41](https://github.com/chat-roulettte/chat-roulette/pull/41)
* `Physical` connection mode renamed to `In Person` in the UI - [#45](https://github.com/chat-roulettte/chat-roulette/pull/45)
* Various enhancements to bot messages - [#46](https://github.com/chat-roulettte/chat-roulette/pull/46)

## v1.2.0 - 2024-08-24

### Fixed
* Channel admin is now the Slack user who adds the chat-roulette bot to the Slack channel rather than the user who created the channel -  [#33](https://github.com/chat-roulettte/chat-roulette/pull/33)

### Added
* Added _Connection Mode_ option -  [#32](https://github.com/chat-roulettte/chat-roulette/pull/32)

### Changed
* App Home now displays the admin of each chat-roulette channel -  [#32](https://github.com/chat-roulettte/chat-roulette/pull/32)

## v1.1.0 - 2024-08-18

### Fixed
* `Monthly` interval is now the same week every month (eg, first Monday of every month) - [#27](https://github.com/chat-roulettte/chat-roulette/pull/27)

### Added
* `QuadWeekly` interval option to preserve the old behavior of `Monthly` - [#27](https://github.com/chat-roulettte/chat-roulette/pull/27)

### Changed
* Jaeger now uses OTLP HTTP exporter - [#28](https://github.com/chat-roulettte/chat-roulette/pull/28)

## v1.0.1 - 2022-12-09

### Fixed
* Removed usage of safehtml to fix History page - [#4](https://github.com/chat-roulettte/chat-roulette/pull/4)
* Updated LinkedIn regex to allow country subdomains - [#6](https://github.com/chat-roulettte/chat-roulette/pull/6)

## v1.0.0 - 2022-11-21
### Added
* initial project release

internal/config/config.go
package config

import (
	"encoding/hex"
	"fmt"
	"os"
	"reflect"
	"strconv"
	"strings"
	"time"

	validation "github.com/go-ozzo/ozzo-validation/v4"
	"github.com/go-ozzo/ozzo-validation/v4/is"
	"github.com/pkg/errors"
	"github.com/spf13/viper"

	"github.com/chat-roulettte/chat-roulette/internal/isx"
)

// Config stores the configuration for the application
type Config struct {
	Bot      SlackBotConfig
	Database DatabaseConfig
	Server   ServerConfig
	Worker   WorkerConfig
	Tracing  TracingConfig
	Dev      bool
}

// DatabaseConfig stores the configuration for using the database
type DatabaseConfig struct {
	// URL is the PostgreSQL connection URL
	//
	// Required
	URL string

	// Connections is the database connections configuration
	//
	// Optional
	Connections DBConnectionsConfig

	// Encryption is the database encryption configuration
	//
	// Required
	Encryption DBEncryptionConfig
}

// DBEncryptionConfig stores the configuration for encrypting data in the database.
type DBEncryptionConfig struct {
	// Key is the current data encryption key (DEK) used to encrypt and decrypt data.
	// This must be a 32-byte hex-encoded key.
	//
	// Required
	Key string

	// PreviousKey is the key used to decrypt old data. This should
	// only be set if the data encryption key is being rotated.
	// If set, this must be a 32-byte hex-encoded key.
	//
	// Optional
	PreviousKey string `mapstructure:"previous_key"`
}

// GetDEK returns the hex-decoded DEK
func (d *DBEncryptionConfig) GetDEK() ([]byte, error) {
	key, err := hex.DecodeString(d.Key)
	if err != nil {
		return nil, errors.Wrap(err, "failed to hex decode data encryption key")
	}

	return key, nil
}

// GetDEK returns the hex-decoded previous DEK
func (d *DBEncryptionConfig) GetPreviousDEK() ([]byte, error) {
	if d.PreviousKey == "" {
		return nil, nil
	}

	key, err := hex.DecodeString(d.PreviousKey)
	if err != nil {
		return nil, errors.Wrap(err, "failed to hex decode previous data encryption key")
	}

	return key, nil
}

// DBConnectionsConfig stores the configuration for the database connection pool
//
// See also: https://go.dev/doc/database/manage-connections
type DBConnectionsConfig struct {
	// MaxOpen is the maximum number of open connections.
	//
	// Optional
	MaxOpen int `mapstructure:"max_open"`

	// MaxIdle is the maximum number of idle connections.
	//
	// Optional
	MaxIdle int `mapstructure:"max_idle"`

	// MaxLifeTime is the maximum lifetime of connections.
	//
	// Optional
	MaxLifetime time.Duration `mapstructure:"max_lifetime"`

	// MaxIdleTime is the maximum amount of time a connection can be idle.
	//
	// Optional
	MaxIdletime time.Duration `mapstructure:"max_idletime"`
}

// ServerConfig stores the configuration for the bot server
type ServerConfig struct {
	// Address is the address that the server binds on.
	//
	// Optional, defaults to 0.0.0.0
	Address string

	// Port is the TCP port that the server binds on.
	//
	// Optional, defaults to 8080
	Port int

	// ClientID is the Slack OpenID Connect (OIDC) client ID.
	//
	// Required
	ClientID string `mapstructure:"client_id"`

	// Client Secret is Slack OpenID Connect (OIDC) client secret.
	//
	// Required
	ClientSecret string `mapstructure:"client_secret"`

	// RedirectURL is the Slack OpenID Connect (OIDC) redirect URL.
	//
	// Required
	RedirectURL string `mapstructure:"redirect_url"`

	// SecretKey is the 32-byte hex-encoded secret key used to authenticate cookies.
	//
	// Required
	SecretKey string `mapstructure:"secret_key"`

	// SigningSecret is the secret used to authenticate requests received from Slack.
	//
	// Required
	SigningSecret string `mapstructure:"signing_secret"`
}

// GetSecretKey returns the hex-decoded Secret Key
func (s *ServerConfig) GetSecretKey() ([]byte, error) {
	key, err := hex.DecodeString(s.SecretKey)
	if err != nil {
		return nil, errors.Wrap(err, "failed to hex decode secret key")
	}

	return key, nil
}

// WorkerConfig stores the configuration for the task queue workers
type WorkerConfig struct {
	// Concurrency is the number of concurrent workers to run
	//
	// Optional, defaults to the number of CPU cores
	Concurrency int
}

// SlackBotConfig stores the configuration for the Slack bot
type SlackBotConfig struct {
	// AuthToken is the Slack OAuth2 bot token
	//
	// Required
	AuthToken string `mapstructure:"auth_token"`
}

// TracingConfig stores the configuration for OpenTelemetry tracing.
// Only one exporter can be configured at a time.
type TracingConfig struct {
	// Enabled turns on OpenTelemetry tracing.
	//
	// Optional
	Enabled bool `mapstructure:"enabled"`

	// Exporter sets the tracing exporter to use.
	// The only options currently supported are: jaeger or honeycomb.
	//
	// Optional
	Exporter TracingExporter `mapstructure:"exporter"`

	// Jaeger stores the configuration for the Jaeger exporter
	//
	// Optional, required if Exporter=jaeger
	Jaeger JaegerTracing `mapstructure:"jaeger"`

	// Honeycomb stores the configuration for the Honeycomb exporter
	//
	// Optional, required if Exporter=honeycomb
	Honeycomb HoneycombTracing `mapstructure:"honeycomb"`
}

// JaegerTracing contains configuration for the Jaeger exporter
type JaegerTracing struct {
	// Endpoint is the URL of the Jaeger collector.
	// This is typically "http://localhost:4318/v1/traces"
	//
	// Required
	Endpoint string `mapstructure:"endpoint"`
}

// HoneycombTracing contains configuration for the Honeycomb exporter
type HoneycombTracing struct {
	// Team is the Honeycomb API key to use
	//
	// Required
	Team string `mapstructure:"team"`

	// Dataset is the destination dataset to send traces to
	//
	// Required
	Dataset string `mapstructure:"dataset"`
}

func newDefaultConfig() *Config {
	return &Config{
		Server: ServerConfig{
			Address: DefaultServerAddr,
			Port:    DefaultServerPort,
		},
		Worker: WorkerConfig{
			Concurrency: DefaultWorkerConcurrency,
		},
		Database: DatabaseConfig{
			Connections: DBConnectionsConfig{
				MaxIdle:     DefaultDBMaxIdle,
				MaxOpen:     DefaultDBMaxOpen,
				MaxLifetime: DefaultDBMaxLifetime,
				MaxIdletime: DefaultDBMaxIdletime,
			},
		},
		Dev: false,
	}
}

// LoadConfig loads configuration from a file and environment variables
func LoadConfig(path string) (*Config, error) {
	v := viper.New()

	// Load default config
	config := newDefaultConfig()

	// Obtain config from environment variables
	v.AutomaticEnv()
	v.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
	bindEnvs(v, config)

	// Obtain config from file, optionally
	if path != "" {
		v.SetConfigFile(path)
		if err := v.ReadInConfig(); err != nil {
			return nil, err
		}
	}

	if err := v.Unmarshal(&config); err != nil {
		return nil, err
	}

	// PaaS typically set the PORT environment variable
	port := os.Getenv("PORT")
	if port != "" {
		v, _ := strconv.Atoi(port)
		config.Server.Port = v
	}

	// Validate the config
	if err := config.Validate(); err != nil {
		return nil, err
	}

	return config, nil
}

// Validate verifies the configuration
func (c Config) Validate() error {
	// Validate bot config
	if err := validation.ValidateStruct(&c.Bot,
		validation.Field(&c.Bot.AuthToken, validation.By(isx.SlackBotAuthToken)),
	); err != nil {
		return errors.Wrap(err, "failed to validate bot config")
	}

	// Validate database config
	if err := validation.ValidateStruct(&c.Database,
		validation.Field(&c.Database.URL, validation.By(isx.PostgresConnectionURL)),
	); err != nil {
		return errors.Wrap(err, "failed to validate database URL")
	}

	if c.Database.Connections.MaxIdle > c.Database.Connections.MaxOpen {
		return fmt.Errorf("max_idle cannot be greater than max_open")
	}

	if err := validation.ValidateStruct(&c.Database.Connections,
		validation.Field(&c.Database.Connections.MaxOpen, validation.Min(5)),
		validation.Field(&c.Database.Connections.MaxIdle, validation.Min(2)),
		validation.Field(&c.Database.Connections.MaxIdletime, validation.Required),
		validation.Field(&c.Database.Connections.MaxLifetime, validation.Required),
	); err != nil {
		return errors.Wrap(err, "failed to validate database config")
	}

	if err := validation.ValidateStruct(&c.Database.Encryption,
		validation.Field(&c.Database.Encryption.Key, validation.Required, is.Hexadecimal, validation.Length(64, 64)),
		validation.Field(&c.Database.Encryption.PreviousKey, is.Hexadecimal, validation.Length(64, 64)),
	); err != nil {
		return errors.Wrap(err, "failed to validate database encryption config")
	}

	// Validate tracing config
	if c.Tracing.Enabled {
		switch c.Tracing.Exporter {
		case TracingExporterJaeger:
			if err := validation.ValidateStruct(&c.Tracing.Jaeger,
				validation.Field(&c.Tracing.Jaeger.Endpoint, validation.Required, is.URL),
			); err != nil {
				return errors.Wrap(err, "failed to validate Jaeger tracing config")
			}
		case TracingExporterHoneycomb:
			if err := validation.ValidateStruct(&c.Tracing.Honeycomb,
				validation.Field(&c.Tracing.Honeycomb.Team, validation.Required, is.Alphanumeric),
				validation.Field(&c.Tracing.Honeycomb.Dataset, validation.Required),
			); err != nil {
				return errors.Wrap(err, "failed to validate Honeycomb tracing config")
			}
		default:
			return errors.Wrap(fmt.Errorf("exporter can only be set to 'jaeger' or 'honeycomb'"), "failed to validate tracing config")
		}
	}

	// Validate server config
	if c.Server.Port < 0 || c.Server.Port > 65536 {
		return errors.Wrap(fmt.Errorf("invalid server port"), "failed to validate server config")
	}

	if err := validation.ValidateStruct(&c.Server,
		validation.Field(&c.Server.Address, validation.Required, is.Host),
		validation.Field(&c.Server.ClientID, validation.Required),
		validation.Field(&c.Server.ClientSecret, validation.Required),
		validation.Field(&c.Server.RedirectURL, validation.By(isx.RedirectURL)),
		validation.Field(&c.Server.SecretKey, validation.Required, is.Hexadecimal, validation.Length(64, 64)),
		validation.Field(&c.Server.SigningSecret, validation.Required),
	); err != nil {
		return errors.Wrap(err, "failed to validate server config")
	}

	// Validate worker config
	if c.Worker.Concurrency < 1 {
		return fmt.Errorf("worker concurrency cannot be less than 1")
	}

	if err := validation.ValidateStruct(&c.Worker); err != nil {
		return errors.Wrap(err, "failed to validate worker config")
	}

	return nil
}

// GetAddr returns the addr of the server
func (c *Config) GetAddr() string {
	return fmt.Sprintf("%s:%d", c.Server.Address, c.Server.Port)
}

// bindEnvs calls viper.BindEnv() for all fields of a struct
//
// This is a workaround for these viper issues:
// https://github.com/spf13/viper/issues/188
// https://github.com/spf13/viper/issues/761
func bindEnvs(v *viper.Viper, iface interface{}, parts ...string) {
	ifValue := reflect.ValueOf(iface)
	ifType := reflect.TypeOf(iface)

	// Dereference if iface is a pointer
	if ifType.Kind() == reflect.Ptr {
		ifValue = ifValue.Elem()
		ifType = ifType.Elem()
	}

	for i := 0; i < ifType.NumField(); i++ {
		value := ifValue.Field(i)
		t := ifType.Field(i)

		tag, ok := t.Tag.Lookup("mapstructure")
		if !ok {
			tag = strings.ToLower(t.Name)
		}

		switch value.Kind() {
		case reflect.Struct:
			bindEnvs(v, value.Interface(), append(parts, tag)...)
		default:
			v.BindEnv(strings.Join(append(parts, tag), ".")) //nolint:errcheck
		}
	}
}
